hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/command/procedures/ShowTimelineProcedure.scala

```diff
@@ -0,0 +1,381 @@
+
+  private def getTimelineEntries(metaClient: HoodieTableMetaClient,
+                                 limit: Int,
+                                 showArchived: Boolean,
+                                 startTime: String,
+                                 endTime: String): Seq[Row] = {
+
+    val instantInfoMap = buildInstantInfoFromTimeline(metaClient)
+
+    val activeRollbackInfoMap = getRolledBackInstantInfo(metaClient.getActiveTimeline, metaClient)
+    val archivedRollbackInfoMap = if (showArchived) {
+      getRolledBackInstantInfo(metaClient.getArchivedTimeline, metaClient)
+    } else {
+      Map.empty[String, List[String]]
+    }
+
+    val activeEntries = getTimelineEntriesFromTimeline(
+      metaClient.getActiveTimeline, "ACTIVE", metaClient, instantInfoMap, activeRollbackInfoMap, limit, startTime, endTime
+    )
+
+    val finalEntries = if (showArchived) {
+      val archivedEntries = getTimelineEntriesFromTimeline(
+        metaClient.getArchivedTimeline, "ARCHIVED", metaClient, instantInfoMap, archivedRollbackInfoMap, limit, startTime, endTime
+      )
+      val combinedEntries = (activeEntries ++ archivedEntries)
+        .sortWith((a, b) => a.getString(0) > b.getString(0))
+
+      if (startTime.trim.nonEmpty && endTime.trim.nonEmpty) {
+        combinedEntries
+      } else {
+        combinedEntries.take(limit)
+      }
+    } else {
+      if (startTime.trim.nonEmpty && endTime.trim.nonEmpty) {
+        activeEntries
+      } else {
+        activeEntries.take(limit)
+      }
+    }
+
+    finalEntries
+  }
+
+  private def getTimelineEntriesFromTimeline(timeline: HoodieTimeline,
+                                             timelineType: String,
+                                             metaClient: HoodieTableMetaClient,
+                                             instantInfoMap: Map[String, Map[HoodieInstant.State, HoodieInstantWithModTime]],
+                                             rollbackInfoMap: Map[String, List[String]],
+                                             limit: Int,
+                                             startTime: String,
+                                             endTime: String): Seq[Row] = {
+
+    val instants = timeline.getInstants.iterator().asScala.toSeq
+
+    val filteredInstants = if (startTime.trim.nonEmpty || endTime.trim.nonEmpty) {
+      instants.filter { instant =>
+        val instantTime = instant.requestedTime()
+        val withinStartTime = if (startTime.trim.nonEmpty) instantTime >= startTime else true
+        val withinEndTime = if (endTime.trim.nonEmpty) instantTime <= endTime else true
+        withinStartTime && withinEndTime
+      }
+    } else {
+      instants
+    }
+
+    val sortedInstants = filteredInstants.sortWith((a, b) => a.requestedTime() > b.requestedTime())
+
+    val limitedInstants = if (startTime.trim.nonEmpty && endTime.trim.nonEmpty) {
+      sortedInstants
+    } else {
+      sortedInstants.take(limit)
+    }
```

Comment:
Why does the limit only take effect when the user does not set the startTime and endTime?